sourced/ml/extractors/type_children.py

```diff
@@ -0,0 +1,122 @@
+import numpy
+import re
+from collections import defaultdict
+
+from sourced.ml.algorithms import UastNodes2Bag
+from sourced.ml.extractors import BagsExtractor, register_extractor, get_names_from_kwargs,\
+    filter_kwargs
+
+
+@register_extractor
+class TypeChildrenBagExtractor(BagsExtractor):
+    NAME = "typechildren"
+    NAMESPACE = "tc."
+    OPTS = dict(get_names_from_kwargs(UastNodes2Bag.__init__))
+    OPTS = BagsExtractor.OPTS.copy()
+
+    def __init__(self, docfreq_threshold=None, quantization=True, **kwargs):
+        super().__init__(docfreq_threshold, quantization)
+        self._log.debug("__init__ %s", kwargs)
+        self.mapping = dict()
+        uast2bag_kwargs = filter_kwargs(kwargs, UastNodes2Bag.__init__)
+        self.uast2bag = UastNodes2Bag(**uast2bag_kwargs)
+
+    def inspect(self, uast):
+        try:
+            bag, all_children = self.uast_to_bag(uast)
+        except RuntimeError as e:
+            raise ValueError(str(uast)) from e
+        if self.quantization:
+            for nb_children in all_children:
+                yield str(nb_children)
+        else:
+            for key in bag:
+                yield self.NAMESPACE + self.quantize(key)
+
+    def quantize(self, key):
+        children_quant = self.mapping[key.split("_")[-1]]
+        return re.sub(r"\d{1,}", str(children_quant), key)
+
+    def extract(self, uast):
+        ndocs = self.ndocs
+        docfreq = self.docfreq
+        log = numpy.log
+        bag, _ = self.uast_to_bag(uast)
+        bag = self.merge_children(bag)
+        for key, val in bag.items():
+            key = self.NAMESPACE + key
+            try:
+                yield key, log(1 + val) * log(ndocs / docfreq[key]) * self.scale
+            except KeyError:
+                # docfreq_threshold
+                continue
+
+    def get_children_freq(self, reduced):
+        children_freq = {}
+        NDOCS_KEY = -1, 0
+        for (i, key), value in reduced.items():
+            if (i, key) != NDOCS_KEY:
+                children_freq[key] = value
+        return children_freq
+
+    def merge_children(self, bag):
+        new_bag = defaultdict(int)
+        for key, val in bag.items():
+            new_bag[self.quantize(key)] += val
+        return new_bag
+
+    def build_quantization(self, children_freq):
+        partition = self.build_partition(children_freq)
+        for value in set(children_freq):
+            self.mapping[value] = self.process_value(int(value), partition)
+        self.quantization = False
+
+    def build_partition(self, children_freq, nb_partitions=15):
+        """
+        Builds the partition of the quantization.
```

Comment:
Please make better explanation. I don't understand what this function is doing and why it's needed